### ISSUE TITLE
Generalized differentiable-Jacobian helper (beorn.differentiate)

### ISSUE THREAD

**Issue by sambit-giri** [cosmic-reionization/BEoRN#51]
## Problem

A downstream analysis script (private, outside this package — differentiable-21cm paper repo) hand-rolls the same pattern three separate times, for three different observables — matter P(k), the halo mass function, and halo P(k):

```python
def theta_vec(theta): ...      # pack a theta dict into a fixed-order jnp array
def _theta_dict(vec): ...      # unpack it back

def matter_power_spectrum_jacobian(theta, noise, L, z, kbins=15, ...):
    def S(vec):
        th = _theta_dict(vec)
        delta = matter_density_field(th, noise, L, z, ...)
        ps, _ = t2c.power_spectrum_1d(1.0 + delta, kbins=kbins, box_dims=L, ...)
        return ps
    ks, ps = matter_power_spectrum(theta, noise, L, z, kbins=kbins, ...)
    J = jax.jacfwd(S)(theta_vec(theta))
    return ks, ps, np.asarray(J)

# hmf_jacobian and halo_power_spectrum_jacobian repeat the exact same shape
```

None of this boilerplate is specific to power spectra or mass functions — it's generic to "differentiate any theta-dependent observable w.r.t. a named parameter vector," and belongs in BEoRN itself rather than being reimplemented per notebook/per observable. This is also directly load-bearing for the differentiable-21cm paper's actual endpoint (exact Fisher forecasts), so getting the shared abstraction right now pays off directly there.

## Proposed design: `beorn.differentiate`

A new top-level module (parallel to the existing `beorn.power_spectrum`, added for #48):

```python
def theta_vector(theta: dict, names: list[str]) -> jnp.ndarray:
    """Pack a parameter dict into a fixed-order vector."""

def theta_dict(vec, names: list[str]) -> dict:
    """Unpack a vector back into a named dict."""

def jacobian(observable_fn, theta, *args, names=None, mode='fwd', backend='jax', **kwargs):
    """d(observable_fn(theta, *args, **kwargs)) / d(theta[names]).

    observable_fn(theta_dict, *args, **kwargs) -> array. mode: 'fwd' (jacfwd,
    cheap for few parameters/many outputs -- the common case here) or 'rev'
    (jacrev/torch.autograd, cheap for many parameters/few outputs -- a
    larger Fisher forecast with more free parameters than output bins).
    Returns (value, jacobian).
    """

def log_derivative(value, jac, theta, names) -> jnp.ndarray:
    """(theta_i / value) * jac[..., i] -- the dln(obs)/dln(theta_i) pattern
    every notebook cell currently re-derives by hand."""
```

Each existing observable (`matter_power_spectrum`, `halo_power_spectrum`, HMF) stays a plain function of `theta` — no bespoke `*_jacobian` twin needed; callers do `beorn.differentiate.jacobian(matter_power_spectrum, theta, param, z, ...)`.

## Two design constraints from recent Parameters/backend consolidation (#48 follow-up, 6dd7beb)

**1. Observable functions this wraps should be Parameters-first, matching `get_density`'s own convention.**
`mass_assignment`/`deconvolve_mas`/`oversample`/`backend`/`use_hunits` are now all `Parameters.simulation` defaults, resolved via a consistent `None`-means-"read from `parameters`" pattern (e.g. `LPTBase.get_density(z, mass_assignment=None, oversample=None, backend=None, deconvolve=None)`). The observable functions this module wraps should follow the same shape — `observable_fn(theta, param, z, ...)` reading painting/oversample/deconvolve/use_hunits defaults off `param`, explicit override still possible — rather than the free-floating-kwargs style the notebook prototype currently uses (`mas=`, `deconvolve_mas=`, `backend=`, no `param` at all).

**2. `jacobian()`'s `backend` argument is a different axis from `Parameters.simulation.backend`, and must not be conflated with it.**
`BackendParameters` (`src/beorn/structs/parameters.py:224`) is explicitly documented as a *production-pipeline performance knob* — "does not make [...] differentiable in the sense of the `beorn.*.differentiable` module family," and jax/torch's scatter-add mass assignment isn't even bit-deterministic on GPU. `jacobian()`'s own `backend` argument selects the *autodiff engine* (`jax.jacfwd`/`jacrev` vs `torch.autograd`) for the `beorn.*.differentiable` function family specifically (`lpt_ics`/`lpt_density`, `halo_field_diff`, `dndlnm_diff`, `pk_eh_nowiggle` — all already support both engines per `beorn.mass_function.models`'s own docstring) — it is unrelated to, and must never read from, `param.simulation.backend`. Worth flagging prominently in the docstring/implementation so this doesn't get silently conflated later.

## Dependency

Blocked-by/related: #49 (its output-unit follow-up comment specifically — P(k)/dn-dlnM don't yet respect `use_hunits`) — any Jacobian this module takes over those observables inherits whatever unit convention their raw output currently has. Sequence that fix first, or at minimum keep this module unit-agnostic (operate on whatever the wrapped `observable_fn` returns, without assuming a unit convention itself) so that fix landing later doesn't require touching this module.

## Explicitly out of scope (future work)

A registry/dataclass bundling `(name, observable_fn, covariance)` per summary statistic, toward assembling an actual Fisher matrix (`Σ_k J^T C⁻¹ J` across several observables) — the paper's actual endpoint, but a separate, larger design decision. This issue is scoped to the generic single-observable Jacobian utility only.

## References

- Prototype to generalize: the `theta_vec`/`_theta_dict`/`matter_power_spectrum_jacobian`/`hmf_jacobian`/`halo_power_spectrum_jacobian` pattern shown above (from a private downstream analysis script, differentiable-21cm paper repo — not part of this package, quoted here in full so no external access is needed).
- #48 follow-up (6dd7beb) — `Parameters.simulation.backend`/`mass_assignment`/`deconvolve_mas`/`oversample` consolidation this design builds on.
- #49 — output-unit dependency (see follow-up comment there).
- `src/beorn/mass_function/models.py` — existing jax/torch dual-backend differentiable primitives this module wraps, doesn't replace.
